### DISCUSSION TITLE
How  to install a theia plugin like "xxx.theia"

### DISCUSSION THREAD

**Discussion by CenGinHong** [eclipse-theia/theia#10076] [Q&A] [answered]
Hi, all
I follow the docs https://theia-ide.org/docs/authoring_plugins/ write a theia plugin, and build it as "xxx.theia".but error would be throw when building if I write it into package.json 
`    
"theiaPlugins": {
             "xxx": "http://xxxxx/xxx.theia"
    }
`

I want to run theia in docker and install the plugin into it，is there any way like 
 `
/usr/bin/code-server --install-extension ms-python.python && \
`
can do that?

BTW ,does xxxx.theia is equal to xxxx.vsix?,can I rename its extname and install it?

**Comment by vince-fugnitto** [comment, marked as answer]
@ChenChienhang thank you for the discussion :+1: 

I'm not sure I follow your question, would you mind clarifying a bit?

> but error would be throw when building if I write it into package.json

The plugin should already be built and packaged before it can be referenced in the application's `package.json`.
The entry `theiaPlugins` is simply used to download already packaged plugins. Do you mind sharing the error you have?

> I want to run theia in docker and install the plugin into it，is there any way like /usr/bin/code-server --install-extension ms-python.python && \ can do that?

You would add it as part of `theiaPlugins`, or any other command which pulls it into the `plugins` folder you specified.

> BTW ,does xxxx.theia is equal to xxxx.vsix?,can I rename its extname and install it?

The `xxxx` part can be whatever name you want. The supported file extensions can be seen from the `download-plugins` script if you use `theiaPlugins`:

https://github.com/eclipse-theia/theia/blob/10d5275207879334d0052a2e8a4d85c82fe30245/dev-packages/cli/src/download-plugins.ts#L161-L168

The script should warn when an unsupported file type is used.

**Reply by CenGinHong** [reply]
I'm so sorry that I didn't express my meaning clearly since I'm not a native English speaker

> but error would be throw when building if I write it into package.json

it means that after the plugin already be bulit and packaged ,it generate a plugins like "xxx.theia" . and then I add it into the package.json of theia and build it by docker,some error will be throw during building because download-plugins don't support "xxx.theia".
so is there any way can I install my plugin into theia-docker?

**Reply by vince-fugnitto** [reply]
I believe you are experiencing https://github.com/eclipse-theia/theia/issues/9264.
You should be able to package your extension as a `.vsix` by performing `vsce package` on your plugin similarly to how vscode plugins are packaged, or in the other supported formats.

**Reply by CenGinHong** [reply]
Thanks a lot, that is what I experiencing, so I'm a little confused if it don't support `.theia`, why the `@theia/generator-plugin` would package it as `.theia` but not `.vsix`.
